k2/csrc/determinize.cc

```diff
@@ -5,469 +5,287 @@
+  // TODO(dpovey): maybe remove this, for efficiency?
+  std::reverse(deriv_out->begin(), deriv_out->end());
 }
 
-void DetState::CheckElementOrder(const Fsa &input_fsa) const {
-  // Checks that the DetStateElements are in a lexicographical order on the
-  // lists of states in their paths.  This will be true becase of how we
-  // construct them (it requires on the IsArcSorted() property, whereby arcs
-  // leaving each state in the FSA are sorted first on label and then on
-  // dest_state.
-  if (seq_len == 0) {
-    CHECK_EQ(elements.size(), 1);
-    CHECK_EQ(elements.front().weight, 0.0);
-  }
-
-  std::vector<int32> prev_seq;
-  for (auto iter = elements.begin(); iter != elements.end(); ++iter) {
-    auto path = iter->path;
-    std::vector<int32> cur_seq;
-    for (int32_t i = 0; i < seq_len; i++) {
-      cur_seq.push_back(input_fsa.arcs[path->arc_index].prev_state);
-      path = path->prev;
-    }
-    std::reverse(cur_seq.begin(), cur_seq.end());
-    if (iter != elements.begin()) {
-      CHECK(cur_seq > prev_seq);
-    }
-    prev_seq.swap(cur_seq);
+void TraceBack(std::unordered_set<MaxTracebackState *> *cur_states,
+               int32_t num_steps,
+               const float *,  // arc_weights_in, unused.
+               float *weight_out, std::vector<int32_t> *deriv_out) {
+  CHECK_EQ(cur_states->size(), 1);
+  MaxTracebackState *state = *(cur_states->begin());
+  double cur_forward_prob = state->forward_prob;
+  deriv_out->resize(num_steps);
+  for (int32_t i = num_steps - 1; i >= 0; i--) {
+    // `deriv_out` is just a list of arc indexes in the input FSA
+    // that this output arc depends on (it's their sum).
+    (*deriv_out)[i] = state->arc_id;
   }
+  double prev_forward_prob = state->forward_prob;
+  *weight_out = cur_forward_prob - prev_forward_prob;
 }
 
-/*
-  This class maps from determinized states (DetState) to integer state-id
-  in the determinized output.
- */
-class DetStateMap {
- public:
-  /*
-    Outputs the output state-id corresponding to a specific DetState structure.
-    This does not store any pointers to the DetState or its contents, so
-    you can delete the DetState without affecting this object's ability to map
-    an equivalent DetState to the same state-id.
+template <class TracebackState>
+bool NormalizeStateAndOutputArc(
+    DetState<TracebackState> *state, const WfsaWithFbWeights &wfsa_in,
+    float prune_cutoff, std::vector<Arc> *arcs_out,
+    std::vector<float> *arc_weights_out,
+    std::vector<std::vector<typename TracebackState::DerivType>>
+        *derivs_per_arc,
+    DetStateMap<TracebackState> *state_map) {
+  float arc_weight;
+  std::vector<typename TracebackState::DerivType> deriv_info;
+  state->Normalize(wfsa_in, &arc_weight, &deriv_info);
+  int32_t next_state_id;
+  bool is_new_state = state_map->GetOutputState(state);
+  arcs_out->push_back({this->state_id, next_state_id, label});
+  arc_weights_out->push_back(arc_weight);
```

Comment:
What does `this` point to? and what's the `next_state_id` , seems we need to change the signature of `GetOutputState` to return a pair of `<bool, next_state_id>`.

And what's the `label`